Create and sign transaction, create multisig address

LeeSmet opened the pull request.

GlenDC commented on modules/consensus/outputs.go:
```diff
@@ -0,0 +1,30 @@
+	"github.com/rivine/rivine/types"
+)
+
+// GetCoinOutput returns the coin output for the given ID
```
returns the **unspent** coin output

GlenDC commented on modules/consensus/outputs.go:
```diff
@@ -0,0 +1,30 @@
+	return co, err
+}
+
+// GetBlockStakeOutput returns the blockstake output for the given ID
```
returns the **unspent** blockstake output

GlenDC commented on modules/wallet/transactionbuilder.go:
```diff
@@ -413,6 +414,136 @@ func (tb *transactionBuilder) Sign() ([]types.Transaction, error) {
 }
 
+// AttemptSigning tries to sign any input for which keys are loaded in the wallet
+func (tb *transactionBuilder) AttemptSigning() error {
```
Why is this called `AttemptSigning`, and not just `Sign`? In the end we always try no, why is it special this time?

LeeSmet replied:
The transactionbuilder already has a method `sign`. This method only works on inputs which have been added with `Fundxxx`, since those inputs are local to the wallet. This also means that no special handling is needed there as they only ever sign the singlesig fulfillments. So I thought it would be best to keep this as a separate method. And in this method, we attempt to find a suitable key to sign every input

GlenDC replied:
Perhaps call it than `SignAllPossibleInputs`?

GlenDC commented on modules/wallet/transactionbuilder.go:
```diff
@@ -413,6 +414,136 @@ func (tb *transactionBuilder) Sign() ([]types.Transaction, error) {
+
+	// Use an inner function so we can drill down into wrapper conditions
+	var f func(int, *types.CoinInput, types.MarshalableUnlockCondition) error
+	var g func(int, *types.BlockStakeInput, types.MarshalableUnlockCondition) error
```
why make inline functions of it, rather than private functions of the `*transactionBuilder`?

LeeSmet replied:
Started out as an anonymous function, which didn't work in the end due to the recursion. And I didn't change it at the time as its only used by the enclosing method

GlenDC replied:
Well I would prefer to turn those into private methods, makes the core function much smaller and readable

GlenDC commented on modules/wallet/wallet_test.go:
```diff
@@ -586,3 +586,29 @@ func (css *consensusSetStub) ConsensusSetSubscribe(subscriber modules.ConsensusS
 }
+
+func (css *consensusSetStub) GetCoinOutput(id types.CoinOutputID) (co types.CoinOutput, err error) {
+	for _, block := range css.blocks {
```
Isn't there a faster way? Given that the consensus contains _all_ blocks with unspent blocks, this would make it quite a slow operation, no?

LeeSmet replied:
As this is implemented on the consensussetstub, the list of blocks and transactions will never be too big imo. Any decent size test should use a real cs. If there is ever a time where this stub is used for thousands of blocks, it will need to be refactored entirely imo

GlenDC replied:
Yep, you're correct, sorry, didn't see that.

GlenDC commented on modules/wallet/wallet_test.go:
```diff
@@ -586,3 +586,29 @@ func (css *consensusSetStub) ConsensusSetSubscribe(subscriber modules.ConsensusS
+}
+
+func (css *consensusSetStub) GetBlockStakeOutput(id types.BlockStakeOutputID) (bso types.BlockStakeOutput, err error) {
+	for _, block := range css.blocks {
```
Same comment as for the `GetCoinOutput` method implementation of above.

LeeSmet replied:
same as above

GlenDC commented on pkg/client/defaultclient.go:
```diff
@@ -185,6 +187,10 @@ func DefaultCLIClient(cfg Config) {
 
+	walletCreateCmd.AddCommand(
+		walletCreateMultisisgAddress,
+		walletCreateCoinTxnCmd)
```
Once we agree on this PR, you could also already copy the code and make it work for the `walletCreateBlockstakeTxnCmd`, seems like small effort compared to the rest of the work done for this PR.

GlenDC commented on pkg/client/walletcmd.go:
```diff
@@ -705,5 +743,86 @@ func walletlistlocked() {
+}
+
+func walletcreatemultisigaddress(cmd *cobra.Command, args []string) {
+	if len(args) < 3 {
```
no need to do this here, you can have that requirement right in the definition of your (cobra) command. See https://godoc.org/github.com/spf13/cobra#MinimumNArgs

GlenDC commented:
Would be awesome should you be able to wrap up this PR today, as well as start+finish issue https://github.com/rivine/rivine/issues/341. If we can get these finished today, I could play a bit with multisig tonight and this weekend at home, using our high level CLI functionality.

GlenDC commented:
Reported the failed CI test as issue #353, do not worry about it @LeeSmet.

GlenDC commented on api/wallet.go:
```diff
@@ -630,9 +632,22 @@ func (api *API) walletCreateCoinTransactionHandler(w http.ResponseWriter, req *h
+func (api *API) walletCreateBlockStakeTransactionHandler(w http.ResponseWriter, req *http.Request, _ httprouter.Params) {
+	var body WalletCreateBlockStakeTransactionPOST
+	if err := json.NewDecoder(req.Body).Decode(&body); err != nil {
+		WriteError(w, Error{"error decoding the supplied coin inputs and outputs: " + err.Error()}, http.StatusBadRequest)
```
`block inputs and outputs`*

GlenDC commented on api/wallet.go:
```diff
@@ -630,9 +632,22 @@ func (api *API) walletCreateCoinTransactionHandler(w http.ResponseWriter, req *h
+	}
+	tx, err := api.wallet.CreateRawTransaction(nil, body.Inputs, nil, body.Outputs, nil)
+	if err != nil {
+		WriteError(w, Error{"error after call to /wallet/create/cointransaction: " + err.Error()}, http.StatusBadRequest)
```
`/wallet/create/blockstaketransaction`

GlenDC commented on modules/wallet/transactionbuilder.go:
```diff
@@ -415,132 +415,130 @@ func (tb *transactionBuilder) Sign() ([]types.Transaction, error) {
+	case *types.MultiSignatureCondition:
+		for _, uh := range condition.UnlockHashes {
+			if key, exists := tb.wallet.keys[uh]; exists {
 				if bsi.Fulfillment.Fulfillment == nil || bsi.Fulfillment.Fulfillment.Equal(nil) {
```
A simpler check would be `bsi.Fulfillment.FulfillmentType() == FulfillmentTypeNil`.

GlenDC commented on pkg/client/walletcmd.go:
```diff
@@ -468,7 +479,7 @@ type outputPair struct {
 }
 
-func parsePairedOutputs(args []string) (pairs []outputPair, err error) {
+func parsePairedOutputs(args []string, blockstake bool) (pairs []outputPair, err error) {
```
personally I would give it a functor `func(string) (types.Currency, error)` in a case like this, would avoid the weird if checking every iterator.

The `ParseCoinString` method already fits, the std ParseUint`, could be wrapped by a private function to do so as well.

However, it's a CLI tool, so I'm less worried about it here, even though imho it does make stuff a bit confusing, as such boolean check makes me always think that it might not be true for all iterations, hence why a boolean was chosen for this.

GlenDC commented on pkg/client/walletcmd.go:
```diff
@@ -808,13 +822,57 @@ func walletcreatecointxn(cmd *cobra.Command, args []string) {
 	if err != nil {
 		Die("Could not create raw transaction from inputs and outputs: ", err)
 	}
-	var resp api.WalletCreateCoinTransactionRESP
```
Don't do that, please do keep it wrapped in a response structured. It might seem silly, but it gives the option in the future to add some extra property to the response, without breaking anything, while that isn't easily possible when returning a pure value.

GlenDC commented on pkg/client/walletcmd.go:
```diff
@@ -808,13 +822,57 @@ func walletcreatecointxn(cmd *cobra.Command, args []string) {
 	}
 
-	json.NewEncoder(os.Stdout).Encode(resp.Transaction)
+	json.NewEncoder(os.Stdout).Encode(resp)
```
This error shouldn't be ignored I think.

LeeSmet replied:
At this point the response has been encoded in the daemon endpoint, and decoded again by the cli,  so an error here would likely be caused by an error in the underlying `Write` call to `stdout`. So can we do something useful with this error then?

GlenDC commented on pkg/client/walletcmd.go:
```diff
@@ -808,13 +822,57 @@ func walletcreatecointxn(cmd *cobra.Command, args []string) {
+		Die("Failed to create transaction:", err)
+	}
+
+	json.NewEncoder(os.Stdout).Encode(resp)
```
This error shouldn't be ignored I think.

GlenDC commented:
@leesmet: I already took part of your code and merged it into master, as to already be able to get the unspent outputs from the consensus. Please rebase your commits so that your branch has no conflicts.

The commit, which contains some of your original code, can be seen at master on https://github.com/rivine/rivine/commit/39def46a633351fda132a2e8b12b7c079f73bf7a

LeeSmet commented:
Rebased and updated, minus the "ignored error" of the json encoding in the cli, as I am not sure if we can do anything useful with the error should it occur. The json has already been encoded and decoded, so an error is likely form the underlying `Write` call, which would indicate a problem with writing to the output pipe.

GlenDC commented on modules/wallet/transactionbuilder.go:
```diff
@@ -413,6 +414,134 @@ func (tb *transactionBuilder) Sign() ([]types.Transaction, error) {
+		}
+	case *types.TimeLockCondition:
+		tb.signCoinInput(idx, ci, condition.Condition)
+	default:
```
You are not supporting *types.NilCondition` yet, while this can also be automatically signed.

GlenDC commented on modules/wallet/transactionbuilder.go:
```diff
@@ -413,6 +414,134 @@ func (tb *transactionBuilder) Sign() ([]types.Transaction, error) {
+		}
+	case *types.TimeLockCondition:
+		tb.signBlockStakeInput(idx, bsi, condition.Condition)
+	default:
```
You are not supporting *types.NilCondition` yet, while this can also be automatically signed.

GlenDC commented on pkg/client/walletcmd.go:
```diff
@@ -430,7 +479,19 @@ type outputPair struct {
+	return _CurrencyConvertor.ParseCoinString(input)
+}
+
+func stringToBlockStake(input string) (types.Currency, error) {
```
`stringToBlockstakes`*

GlenDC commented on pkg/client/walletcmd.go:
```diff
@@ -430,7 +479,19 @@ type outputPair struct {
+// parseCurrencyString takes the string representation of a currency value
+type parseCurrencyString func(string) (types.Currency, error)
+
+func stringToCoins(input string) (types.Currency, error) {
```
No need for this function, yo can just use `_CurrencyConvertor.ParseCoinString` directly.

GlenDC commented on modules/wallet/transactionbuilder.go:
```diff
@@ -413,6 +414,174 @@ func (tb *transactionBuilder) Sign() ([]types.Transaction, error) {
+			tb.signed = true
+		}
+	case *types.TimeLockCondition:
+		tb.signCoinInput(idx, ci, condition.Condition)
```
You forgot a `return` keyword here, no?

GlenDC commented on modules/wallet/transactionbuilder.go:
```diff
@@ -413,6 +414,174 @@ func (tb *transactionBuilder) Sign() ([]types.Transaction, error) {
+			tb.signed = true
+		}
+	case *types.TimeLockCondition:
+		tb.signBlockStakeInput(idx, bsi, condition.Condition)
```
You forgot a `return` keyword here, no?

GlenDC commented on modules/wallet/transactionbuilder.go:
```diff
@@ -413,6 +414,174 @@ func (tb *transactionBuilder) Sign() ([]types.Transaction, error) {
+		// Sign using a random key, if any is present
+		if len(tb.wallet.keys) > 0 {
+			var key spendableKey
+			for _, key = range tb.wallet.keys {
```
This doesn't get a crypto-random key though. Either get a crypto-random key, or use perhaps just `getAddress`, which is meant for those purposes.
